lib/Dialect/SV/Transforms/HWLegalizeModules.cpp

```diff
@@ -123,12 +296,54 @@ Operation *HWLegalizeModulesPass::tryLoweringArrayGet(hw::ArrayGetOp getOp) {
         return thePattern;
       });
 
-  // Ok, emit the read from the wire to get the value out.
-  builder.setInsertionPoint(getOp);
-  auto readWire = builder.create<sv::ReadInOutOp>(getOp.getLoc(), theWire);
-  getOp.getResult().replaceAllUsesWith(readWire);
-  getOp->erase();
-  return readWire;
+  return theWire;
+}
+
+bool HWLegalizeModulesPass::processUsers(Operation &op, Value value,
+                                         ArrayRef<Value> mapping) {
+  for (auto *user : llvm::make_early_inc_range(value.getUsers())) {
+    if (TypeSwitch<Operation *, bool>(user)
+            .Case<hw::ArrayGetOp>([&](hw::ArrayGetOp getOp) {
+              if (auto indexAndBitWidth =
+                      tryExtractIndexAndBitWidth(getOp.getIndex())) {
+                getOp.replaceAllUsesWith(mapping[indexAndBitWidth->first]);
+                return true;
+              }
+
+              return false;
+            })
+            .Case<sv::ArrayIndexInOutOp>([&](sv::ArrayIndexInOutOp indexOp) {
+              if (auto indexAndBitWidth =
+                      tryExtractIndexAndBitWidth(indexOp.getIndex())) {
+                indexOp.replaceAllUsesWith(mapping[indexAndBitWidth->first]);
+                return true;
+              }
+
+              return false;
+            })
+            .Default([](auto op) { return false; })) {
+      user->erase();
+      continue;
+    }
+
+    user->emitError("unsupported packed array expression");
+    signalPassFailure();
+    return false;
+  }
+
+  return true;
+}
+
+std::optional<std::pair<size_t, unsigned>>
+HWLegalizeModulesPass::tryExtractIndexAndBitWidth(Value value) {
+  if (auto constantOp = dyn_cast<hw::ConstantOp>(value.getDefiningOp())) {
+    auto index = constantOp.getValue();
+    assert(index.getBitWidth() <= sizeof(size_t) * 8 &&
```

Comment:
We don't want to make assertion behaivor environment dependent. Can we change the returned type to uint64_t or uint32_t?